Add blebox binary_sensor platform

riokuu opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Added `binary_sensor` platform to complement need for integration with rain sensor. Which is a part of multiSensor API(blebox_uniapi).


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
  - https://github.com/home-assistant/core/pull/76295
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/24635

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

riokuu committed:
Add binary_sensor platform, with test.

homeassistant commented:
Hey there @bbx-a, mind taking a look at this pull request as it has been labeled with an integration (`blebox`) you are listed as a [code owner](https://github.com/home-assistant/core/blob/dev/CODEOWNERS#L140) for? Thanks!

epenet commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,38 @@
+    @property
+    def device_class(self):
+        """Return the device class."""
+        return self._feature.device_class
```
Normally we require components to have a static mapping between "component device class" and "home assistant device class"

For example:
```python
BLEEBOX_TO_HASS_DEVICE_CLASS : dict[str, BinarySensorDeviceClass] = {"moisture": BinarySensorDeviceClass.MOISTURE}
```

and then

```python
        return BLEEBOX_TO_HASS_DEVICE_CLASS[self._feature.device_class]
```

I see that this isn't applied in the existing code so maybe there is a reason. I am just pointing it out.

riokuu replied:
Agree, I'll be using class based of `EntityDescription` to fulfill static mapping requirements. 
This way I'll be coherent with `sensor` platform.

epenet commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,38 @@
+"""BleBox sensor entities."""
+
+from homeassistant.components.binary_sensor import BinarySensorEntity
```
```suggestion
from blebox_uniapi.binary_sensor import BinarySensor as BinarySensorFeature

from homeassistant.components.binary_sensor import BinarySensorEntity
```

epenet commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,38 @@
+
+
+class BleBoxBinarySensorEntity(BleBoxEntity, BinarySensorEntity):
+    """Representation of a BleBox sensor feature."""
+
```
```suggestion
    """Representation of a BleBox sensor feature."""

    _feature: BinarySensorFeature

```

epenet commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,38 @@
+        return self._feature.state
+
+    @property
+    def device_class(self):
```
```suggestion
    def device_class(self) -> BinarySensorDeviceClass:
```

riokuu replied:
Method removed, EntityDescriptor used instead.

epenet commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,38 @@
+    """Representation of a BleBox sensor feature."""
+
+    @property
+    def is_on(self):
```
```suggestion
    def is_on(self) -> bool:
```

epenet reviewed (commented):
I suggest adding full type hints in the binary_sensor module.

I also think some improvements should be made in `blebox_uniapi` to ensure that a `state` property is always available on the base `BinarySensor` feature, not just on the `Rain` feature.

riokuu committed:
Applied suggestions by @epenet

riokuu commented:
@epenet Hi there, is there any action required from me? BR

epenet commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,62 @@
+"""BleBox sensor entities."""
```
```suggestion
"""BleBox binary sensor entities."""
```

epenet commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,62 @@
+from . import BleBoxEntity, create_blebox_entities
+
+
+@dataclass
+class BleboxBinarySensorEntityDescription(BinarySensorEntityDescription):
+    """Class describing Blebox binary sensor entities."""
+
+
```
This class doesn't add anything that `BinarySensorEntityDescription` doesn't already provide.
```suggestion
```

epenet commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,62 @@
+
+
+BINARY_SENSOR_TYPES = (
+    BleboxBinarySensorEntityDescription(
```
```suggestion
    BinarySensorEntityDescription(
```

epenet commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,62 @@
+    """Representation of a BleBox binary sensor feature."""
+
+    def __init__(self, feature: BinarySensorFeature) -> None:
+        """Initialize a BleBox sensor feature."""
```
```suggestion
        """Initialize a BleBox binary sensor feature."""
```

epenet commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,62 @@
+        """Initialize a BleBox sensor feature."""
+        super().__init__(feature)
+
+        for description in BINARY_SENSOR_TYPES:
+            if description.key == feature.device_class:
+                self.entity_description = description
+                break
```
This is bad behavior.
It shouldn't be part of `__init__` and the logic should be moved to `async_setup_entry`

riokuu replied:
@epenet Can you elaborate? As I see in major other integrations, self.entity_description is set on the entity's init. I would like to keep it consistent. 
Your suggestion would only resolve descriptions for `BleBoxBinarySensorEntity`, but there is `BleBoxSensorEntity` which also uses entity_descriptors. Also visible prospect application of entity_descriptort in other blebox platforms f.e. light platform.

epenet replied:
self.entity_description is always set on the entity's init.
BUT it is normally passed as an argument, not looked up during the entity's init.
The lookup should be done BEFORE the entity's init.

epenet replied:
Regarding `BleBoxSensorEntity`, in a follow-up PR I suggest that apply the same logic that I recommended here and remove reference to the central `create_blebox_entities`.
I just don't feel that `create_blebox_entities` method adds much value, and just makes it harder to review the code.

Or maybe add a new helper method so that you share one loop between all platforms, but keep the class instance creating inside each platform:
```
@callback
def get_blebox_features(hass, config_entry, entity_type) -> list[Feature]`:
```

epenet commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,62 @@
+) -> None:
+    """Set up a BleBox entry."""
+
+    create_blebox_entities(
+        hass,
+        config_entry,
+        async_add_entities,
+        BleBoxBinarySensorEntity,
+        "binary_sensors",
+    )
```
I don't think you should delegate this to `__init__`

I think you should instead do something like this, so you can pass the description during initialisation:
```suggestion
    product: Box = hass.data[DOMAIN][config_entry.entry_id][PRODUCT]
    entities = []

    feature: Feature
    if features := product.features.get("binary_sensors"):
        for feature in features:
            for description in BINARY_SENSOR_TYPES:
                if description.key == feature.device_class:
                    entities.append(BleBoxBinarySensorEntity(feature, description))
                    break

    async_add_entities(entities, True)
```

riokuu committed:
refactor: as @epenet suggested, passing entity_description to init

epenet commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,54 @@
+) -> None:
+    """Set up a BleBox entry."""
+
+    create_blebox_entities(
+        hass,
+        config_entry,
+        async_add_entities,
+        BleBoxBinarySensorEntity,
+        "binary_sensors",
+        BINARY_SENSOR_TYPES,
+    )
```
You misunderstood me. I was refering to `BleBoxBinarySensorEntity.__init__()` not refering to `__init__` module.

I think that you should NOT reference `create_blebox_entities`, and instead move the logic here.

```suggestion
    product: Box = hass.data[DOMAIN][config_entry.entry_id][PRODUCT]
    entities: list[BleBoxBinarySensorEntity] = []

    if "binary_sensors" in product.features:
        for feature in product.features["binary_sensors"]:
            for description in BINARY_SENSOR_TYPES:
                if description.key == feature.device_class:
                    entities.append(entity_klass(feature, description))
                    break
    async_add_entities(entities, True)
```

epenet commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -66,17 +67,27 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 @callback
 def create_blebox_entities(
-    hass, config_entry, async_add_entities, entity_klass, entity_type
+    hass,
+    config_entry,
+    async_add_entities,
+    entity_klass,
+    entity_type,
+    entity_descriptors=None,
```
You don't need to adjust `create_blebox_entities` in this PR.
```suggestion
    hass, config_entry, async_add_entities, entity_klass, entity_type
```

epenet commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -66,17 +67,27 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Create entities from a BleBox product's features."""
 
-    product = hass.data[DOMAIN][config_entry.entry_id][PRODUCT]
+    product: Box = hass.data[DOMAIN][config_entry.entry_id][PRODUCT]
```
You don't need to `adjust create_blebox_entities` in this PR.
```suggestion
    product = hass.data[DOMAIN][config_entry.entry_id][PRODUCT]
```

epenet commented on homeassistant/components/blebox/__init__.py:
```diff
@@ -66,17 +67,27 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
         for feature in product.features[entity_type]:
-            entities.append(entity_klass(feature))
-
+            if entity_descriptors:
+                for description in entity_descriptors:
+                    if description.key == feature.device_class:
+                        entities.append(entity_klass(feature, description))
+                        break
+            else:
+                entities.append(entity_klass(feature))
```
You don't need to adjust `create_blebox_entities` in this PR.
```suggestion
            entities.append(entity_klass(feature))

```

riokuu committed:
Update homeassistant/components/blebox/binary_sensor.py

@epenet suggestion, moved refactored logic of create_blebox_entities into BleBoxBinarySensorEntity init

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

epenet commented:
Please take some time to review and test #80511.
That PR implements TypeVar, and improves type hints.

riokuu committed:
refactor: as @epenet class selector and entity creation moved to binary_sensor

epenet commented:
Let's wait until https://github.com/home-assistant/core/pull/80511 is merged, then you can rebase over that.

epenet commented:
https://github.com/home-assistant/core/pull/80511 has been merged. Can you rebase please?

riokuu commented:
@epenet rebased.

epenet commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,57 @@
+    async_add_entities(entities, True)
+
+
+class BleBoxBinarySensorEntity(BleBoxEntity, BinarySensorEntity):
```
Can you add generic type?
```suggestion
class BleBoxBinarySensorEntity(BleBoxEntity[BinarySensorFeature], BinarySensorEntity):
```

epenet commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,57 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up a BleBox entry."""
+
+    product: Box = hass.data[DOMAIN][config_entry.entry_id][PRODUCT]
+    entities: list[BleBoxBinarySensorEntity] = []
+
+    if "binary_sensors" in product.features:
+        for feature in product.features["binary_sensors"]:
+            for description in BINARY_SENSOR_TYPES:
+                if description.key == feature.device_class:
+                    entities.append(BleBoxBinarySensorEntity(feature, description))
+                    break
+    async_add_entities(entities, True)
```
Can you use list comprehension (untested)?
```suggestion
    product: Box = hass.data[DOMAIN][config_entry.entry_id][PRODUCT]
    entities = [
        BleBoxBinarySensorEntity(feature) 
        for feature in product.features.get("binary_sensors", [])
        for description in BINARY_SENSOR_TYPES
        if description.key == feature.device_class
    ]
    async_add_entities(entities, True)
```

riokuu replied:
Tested, added in a commit.

riokuu committed:
refactor: list comprehension in entity list setup in binary sensor

riokuu commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,55 @@
+    """Set up a BleBox entry."""
+
+    product: Box = hass.data[DOMAIN][config_entry.entry_id][PRODUCT]
+    entities: list[BleBoxBinarySensorEntity] = [
```
Isn't it self-explanatory/over-zeal?

epenet replied:
You are right, it's inferred here. You can use 
```suggestion
    entities = [
```

riokuu committed:
Update homeassistant/components/blebox/binary_sensor.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

epenet commented:
Looks good. Did you prepare a documentation PR?

riokuu commented:
@epenet we are discussing internally to change the whole thing to be more user friendly/product oriented. If it's necessary I'll update it at once.

epenet commented:
It is needed for this PR

riokuu commented:
Link to documentation pull request updated

epenet reviewed (approved):
Thanks @riokuu 👍

epenet commented on homeassistant/components/blebox/binary_sensor.py:
```diff
@@ -0,0 +1,55 @@
+    """Set up a BleBox entry."""
+
+    product: Box = hass.data[DOMAIN][config_entry.entry_id][PRODUCT]
+    entities = [
+        BleBoxBinarySensorEntity(feature, description)
+        for feature in product.features.get("binary_sensors", [])
+        for description in BINARY_SENSOR_TYPES
+        if description.key == feature.device_class
+    ]
```
Note: please open a PR to adjust the other platforms (I think at least `sensor`) to use the same logic.

epenet commented on tests/components/blebox/test_binary_sensor.py:
```diff
@@ -0,0 +1,45 @@
+
+
+@pytest.fixture(name="rainsensor")
+def airsensor_fixture():
```
As you know, I am a big fan of type hints, even in tests.
It was not compulsory for this PR, but it would be great to add here

riokuu replied:
Type hints should be in same PR with sensor description refactor?

epenet replied:
Smaller PRs are easier to review. I suggest to split from sensor description refactor.

epenet commented on tests/components/blebox/test_binary_sensor.py:
```diff
@@ -0,0 +1,45 @@
+    return (feature, "binary_sensor.windrainsensor_0_rain")
+
+
+async def test_init(rainsensor, hass, config):
```
Same here, type hints would be good in follow-up PR.
